google/cloud/spanner/internal/session_pool.cc

```diff
@@ -133,11 +115,99 @@ void SessionPool::ScheduleBackgroundWork(std::chrono::seconds relative_time) {
 }
 
 void SessionPool::DoBackgroundWork() {
+  MaintainPoolSize();
   // TODO(#1171) Implement SessionPool session refresh
-  // TODO(#1172) maintain desired SessionPool size
   ScheduleBackgroundWork(std::chrono::seconds(5));
 }
 
+void SessionPool::MaintainPoolSize() {
+  // Ensure the pool size conforms to what was specified in the
+  // `SessionOptions`, creating or deleting sessions as necessary.
+  std::unique_lock<std::mutex> lk(mu_);
+  if (!create_in_progress_ && total_sessions_ < options_.min_sessions()) {
+    Grow(lk, total_sessions_ - options_.min_sessions(), /*async=*/true);
+  }
+}
+
+/**
+ * Grow the session pool by creating up to `sessions_to_create` sessions and
+ * adding them to the pool.  Note that `lk` may be dropped in the process.
+ *
+ * TODO(#1271) eliminate the `async` parameter and do all creation
+ * asynchronously. The main obstacle is making existing tests pass.
+ */
+Status SessionPool::Grow(std::unique_lock<std::mutex>& lk,
+                         int sessions_to_create, bool async) {
+  create_in_progress_ = true;
```

Comment:
Nit: Could this simply be moved down to avoid the `= false` in the "already at max" case?  It seems to be something that should be closely bound to `lk`.